keras/backend/ops.py

```diff
@@ -0,0 +1,121 @@
+from ..layers import Lambda
+from ..engine.topology import merge
+from .. import backend as K
+
+
+def _slice_shape(shape, slices):
+    '''slice a shape
+    '''
+    output_shape = []
+    if type(slices) in [int, slice]:
+        slices = [slices]
+    while len(shape) > len(slices):
+        slices += (slice(None),)
+    for i in range(len(slices)):
+        s = slices[i]
+        if type(s) == int:
+            continue
+        else:
+            start = s.start
+            stop = s.stop
+            step = s.step
+            if not start:
+                start = 0
+            if not step:
+                step = 1
+            if shape[i] is None:
+                if stop >= 0 and start >= 0:
+                    output_shape.append((stop - start) / step)
+                else:
+                    output_shape.append(None)
+            else:
+                if start < 0:
+                    start = shape[i] + start
+                if not stop:
+                    stop = shape[i]
+                elif stop < 0:
+                    stop = shape[i] + stop
+                output_shape.append((stop - start) / step)
+    for i in range(len(output_shape)):
+        if output_shape[i] and output_shape[i] < 0:
+            output_shape[i] = 0
+    return tuple(output_shape)
+
+def _compatible(shape1, shape2):
+    '''check if shapes of 2 tensors are compatible for addition, multiplication etc.
+    '''
+    if len(shape1) != len(shape2):
+        return False
+    for dim1, dim2 in zip(shape1, shape2):
+        if dim1 and dim2 and dim1 != dim2:
+            return False
+    return True
+
+def _override_operator(tensor_class, operator):
+    _operator = '_keras' + operator[1:-2]  # we add '_keras' so that it does not conflict with any existing functions
+    setattr(tensor_class, _operator, getattr(tensor_class, operator))
+    if operator in ['__neg__', '__pos__', '__abs__', '__invert__']:
+        # unary operator
+        def op(x):
+            x_k = hasattr(x, '_keras_history')
+            if not x_k:
+                return getattr(x, _operator)()
+            else:
+                def func(x):
+                    return getattr(x, _operator)()
+                lambda_layer = Lambda(func, output_shape=lambda x: x)
+                lambda_layer.build(None)
+                lambda_layer.supports_masking = True
+                return lambda_layer(x)
+    else:
```

Comment:
Rather than a catch-all "else", shouldn't we use an operator whitelist instead? Would seem safer.